src/pymor/models/transfer_function.py

```diff
@@ -120,62 +121,109 @@ def eval_dtf(self, s, mu=None):
         else:
             return self.dtf(s, mu=mu)
 
-    def freq_resp(self, w, mu=None):
+    @cached
+    def freq_resp(self, w, mu=None, adaptive_type='bode', adaptive_opts=None):
         """Evaluate the transfer function on the imaginary axis.
 
         Parameters
         ----------
         w
-            A sequence of angular frequencies at which to compute the transfer function.
+            If `len(w) == 2`, the left and right limits used for the adaptive sampling.
+            Otherwise, a sequence of angular frequencies at which to compute the transfer function.
         mu
             |Parameter values| for which to evaluate the transfer function.
+        adaptive_type
+            The plot type that adaptive sampling should be used for
+            (`'bode'`, `'mag'`).
+            Ignored if `len(w) != 2`.
+        adaptive_opts
+            Optional arguments for `pymor.algorithms.plot.adaptive` (ignored if `len(w) != 2`).
 
         Returns
         -------
+        w
+            A sequence of angular frequencies at which the transfer function was computed
+            (returned if `len(w) == 2`).
         tfw
             Transfer function values at frequencies in `w`, |NumPy array| of shape
             `(len(w), self.dim_output, self.dim_input)`.
         """
-        if self.sampling_time > 0 and not all(-np.pi <= wi <= np.pi for wi in w):
-            self.logger.warning('Some frequencies are not in the [-pi, pi] interval.')
-        w = 1j * w if self.sampling_time == 0 else np.exp(1j * w)
         if not isinstance(mu, Mu):
             mu = self.parameters.parse(mu)
         assert self.parameters.assert_compatible(mu)
-        return np.stack([self.eval_tf(wi, mu=mu) for wi in w])
 
-    def bode(self, w, mu=None):
+        if self.sampling_time > 0 and not all(-np.pi <= wi <= np.pi for wi in w):
+            self.logger.warning('Some frequencies are not in the [-pi, pi] interval.')
+
+        if len(w) == 2:
+            if adaptive_type == 'bode':
+                if self.sampling_time == 0:
+                    f = lambda w: self.eval_tf(1j * w, mu=mu)
+                else:
+                    f = lambda w: self.eval_tf(np.exp(1j * w), mu=mu)
+            else:
+                if self.sampling_time == 0:
+                    f = lambda w: spla.norm(self.eval_tf(1j * w, mu=mu))
+                else:
+                    f = lambda w: spla.norm(self.eval_tf(np.exp(1j * w), mu=mu))
+
+            if adaptive_opts is None:
+                adaptive_opts = {}
+
+            w_new, _ = adaptive(f, w[0], w[1], xscale='log', yscale='log', **adaptive_opts)
```

Comment:
Since `xscale` and `yscale` are set to `log` it is not possible to switch to linear by specifying it in `adaptive_opts` (it would cause an error because these arguments are specified multiple times). Maybe it is on purpose, but if it's not the scaling arguments should be appended to `adaptive_opts` or something similar.

Reply:
Yes, it was on purpose, but thinking about it again, I guess there might be transfer functions that look better on a linear scale. Also, if we ever add Nyquist plots, then linear scaling definitely makes sense. I now set the scales to `'log'` if they are not set by the user.